File: workers/event-bus.ts
import {
  validateEncodedAperioEvent,
  type EncodedAperioEvent
} from "@aperio/shared/protobuf-contracts";

type EventPublisher = {
  publish(event: EncodedAperioEvent): Promise<void>;
  close(): Promise<void>;
};

class NoopEventPublisher implements EventPublisher {
  async publish(): Promise<void> {}
  async close(): Promise<void> {}
}

class NatsJetStreamPublisher implements EventPublisher {
  private connectionPromise: Promise<{
    connection: import("nats").NatsConnection;
    jetstream: import("nats").JetStreamClient;
  }> | null = null;

  constructor(
    private readonly servers: string,
    private readonly streamName: string
  ) {}

  private async connect() {
    if (!this.connectionPromise) {
      this.connectionPromise = (async () => {
        const { connect } = await import("nats");
        const connection = await connect({
          servers: this.servers.split(",").map((server) => server.trim())
        });
        await ensureStream(connection, this.streamName);
        return {
          connection,
          jetstream: connection.jetstream()
        };
      })().catch((error) => {
        this.connectionPromise = null;
        throw error;
      });
    }
    return this.connectionPromise;
  }

  async publish(event: EncodedAperioEvent): Promise<void> {
    const { jetstream } = await this.connect();
    await jetstream.publish(event.subject, event.payload, {
      msgID: event.id
    });
  }

  async close(): Promise<void> {
    if (!this.connectionPromise) return;
    const { connection } = await this.connectionPromise;
    await connection.drain();
    this.connectionPromise = null;
  }
}

async function ensureStream(
  connection: import("nats").NatsConnection,
  streamName: string
) {
  const manager = await connection.jetstreamManager();
  try {
    await manager.streams.info(streamName);
  } catch {
    await manager.streams.add({
      name: streamName,
      subjects: ["events.>"]
    });
  }
}

let publisher: EventPublisher | null = null;

export function getEventPublisher(): EventPublisher {
  if (publisher) return publisher;
  const backend = process.env.APERIO_EVENT_BUS?.trim().toLowerCase() ?? "noop";
  if (backend === "nats") {
    const servers = process.env.APERIO_NATS_URL?.trim() || "nats://127.0.0.1:4222";
    const streamName = process.env.APERIO_NATS_STREAM?.trim() || "CEREBRO_EVENTS";
    publisher = new NatsJetStreamPublisher(servers, streamName);
  } else {
    publisher = new NoopEventPublisher();
  }
  return publisher;
}

export async function publishAperioEvent(event: EncodedAperioEvent): Promise<void> {
  try {
    validateEncodedAperioEvent(event);
    await getEventPublisher().publish(event);
  } catch (error) {
    console.warn(
      "event_bus.publish_failed",
      error instanceof Error ? error.message : "unknown error"
    );
  }
}

export async function closeEventPublisher(): Promise<void> {
  if (!publisher) return;
  await publisher.close();
  publisher = null;
}

File: internal/bootstrap/event_bus.go
package bootstrap

import (
	"context"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/nats-io/nats.go"
	aperiocontractsv1 "github.com/writer/aperio/gen/aperio/contracts/v1"
	cerebrov1 "github.com/writer/aperio/gen/cerebro/v1"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
)

const (
	aperioEventSourceID              = "aperio"
	findingLifecycleSchemaRef        = "aperio/finding_lifecycle/v1"
	findingResolvedEventKind         = "aperio.finding.resolved"
	defaultAperioNATSURL             = "nats://127.0.0.1:4222"
	defaultAperioNATSStream          = "CEREBRO_EVENTS"
	aperioEventPublishConnectTimeout = 5 * time.Second
)

type encodedAperioEvent struct {
	id      string
	subject string
	payload []byte
}

type aperioEventBus struct {
	mu         sync.Mutex
	servers    string
	streamName string
	nc         *nats.Conn
	js         nats.JetStreamContext
}

func (a *App) publishFindingLifecycleEvent(ctx context.Context, findingID, organizationID, integrationID, previousStatus, nextStatus, actorUserID, resolutionNote string, occurredAt time.Time) {
	event, err := encodeFindingLifecycleEvent(findingID, organizationID, integrationID, previousStatus, nextStatus, actorUserID, resolutionNote, occurredAt)
	if err != nil {
		return
	}
	_ = a.publishAperioEvent(ctx, event)
}

func encodeFindingLifecycleEvent(findingID, organizationID, integrationID, previousStatus, nextStatus, actorUserID, resolutionNote string, occurredAt time.Time) (encodedAperioEvent, error) {
	eventID := compatID("evt")
	timestamp := timestamppb.New(occurredAt)
	lifecyclePayload, err := proto.Marshal(&aperiocontractsv1.FindingLifecycleEvent{
		FindingId:      findingID,
		OrganizationId: organizationID,
		IntegrationId:  integrationID,
		PreviousStatus: previousStatus,
		NextStatus:     nextStatus,
		ActorUserId:    actorUserID,
		StatusSource:   "user",
		OccurredAt:     timestamp,
		ResolutionNote: resolutionNote,
	})
	if err != nil {
		return encodedAperioEvent{}, err
	}
	kind := findingResolvedEventKind
	envelopePayload, err := proto.Marshal(&cerebrov1.EventEnvelope{
		Id:         eventID,
		TenantId:   organizationID,
		SourceId:   aperioEventSourceID,
		Kind:       kind,
		OccurredAt: timestamp,
		SchemaRef:  findingLifecycleSchemaRef,
		Payload:    lifecyclePayload,
		Attributes: compactEventAttributes(map[string]string{
			"finding_id":      findingID,
			"integration_id":  integrationID,
			"previous_status": previousStatus,
			"next_status":     nextStatus,
			"actor_user_id":   actorUserID,
			"status_source":   "user",
		}),
	})
	if err != nil {
		return encodedAperioEvent{}, err
	}
	return encodedAperioEvent{
		id:      eventID,
		subject: "events." + kind,
		payload: envelopePayload,
	}, nil
}

func (a *App) publishAperioEvent(ctx context.Context, event encodedAperioEvent) error {
	if strings.ToLower(strings.TrimSpace(os.Getenv("APERIO_EVENT_BUS"))) != "nats" {
		return nil
	}
	if a.eventBus == nil {
		a.eventBus = &aperioEventBus{}
	}
	return a.eventBus.publish(ctx, event)
}

func (b *aperioEventBus) publish(ctx context.Context, event encodedAperioEvent) error {
	servers := strings.TrimSpace(os.Getenv("APERIO_NATS_URL"))
	if servers == "" {
		servers = defaultAperioNATSURL
	}
	streamName := strings.TrimSpace(os.Getenv("APERIO_NATS_STREAM"))
	if streamName == "" {
		streamName = defaultAperioNATSStream
	}
	b.mu.Lock()
	defer b.mu.Unlock()
	js, err := b.jetStream(ctx, servers, streamName)
	if err != nil {
		return err
	}
	if _, err := js.StreamInfo(streamName); err != nil {
		_, _ = js.AddStream(&nats.StreamConfig{Name: streamName, Subjects: []string{"events.>"}})
	}
	_, err = js.Publish(event.subject, event.payload, nats.MsgId(event.id), nats.Context(ctx))
	return err
}

func (b *aperioEventBus) jetStream(ctx context.Context, servers string, streamName string) (nats.JetStreamContext, error) {
	if b.nc != nil && !b.nc.IsClosed() && b.servers == servers && b.streamName == streamName && b.js != nil {
		return b.js, nil
	}
	if b.nc != nil {
		b.nc.Close()
	}
	nc, err := nats.Connect(servers, nats.Timeout(aperioEventPublishConnectTimeout))
	if err != nil {
		return nil, err
	}
	js, err := nc.JetStream(nats.Context(ctx))
	if err != nil {
		nc.Close()
		return nil, err
	}
	b.servers = servers
	b.streamName = streamName
	b.nc = nc
	b.js = js
	return js, nil
}

func compactEventAttributes(attributes map[string]string) map[string]string {
	out := map[string]string{}
	for key, value := range attributes {
		if value != "" {
			out[key] = value
		}
	}
	return out
}
